### ISSUE TITLE
0.15 BreachDepressions produces unmodfied output

### ISSUE THREAD

**Issue by spaceof7** [jblindsay/whitebox-tools#29]
When i run breach depressions I get output that is identical to the input.  I tested on Windows and Linux.  0.14 works correctly.

**Comment by jblindsay** [comment]
@spaceof7 There have been [some changes](https://github.com/jblindsay/whitebox-tools/commit/f422faa9432105848b339eacdedb807734fcee1b#diff-d69144c51343db5c8aeb55017d3aefc3) to the BreachDepression tool in 0.15. Unfortunately, I did not document these changes in the readme.txt in the rush to publish the latest version. Essentially, I added the option to fill single-cell pits prior to breaching, and I added an option to fill the DEM post-breaching to ensure that there is complete removal of depressions. This brings the tool more inline with the original paper's algorithm, which was originally implemented in Go as part of GoSpatial. 

Although I did do extensive testing of the tool before publishing the modified form, it is possible that some bug has been incorporated. Would it be possible for you to provide the details of how you ran BreachDepressions? That is, what environment were you using (Python API, QGIS or ArcGIS plugin, R API, the WB Runner GUI)? What parameters did you use? Is it possible to share your input DEM with me?

**Comment by jblindsay** [comment]
I just did another testing of the algorithm and found that it worked as expected in both full-breach and constrained breaching modes.

Original DEM:
<img width="966" alt="Screen Shot 2019-03-10 at 9 34 09 AM" src="https://user-images.githubusercontent.com/8441542/54085769-ef52ca80-4317-11e9-8dbb-d34c9a2534dd.png">

Full-breached DEM:

```
/whitebox_tools --run="Breachdepressions" --wd="/Users/johnlindsay/Documents/data/Rondeau" --dem='/Users/johnlindsay/Documents/data/Rondeau/DEM_clipped.dep' --output='/Users/johnlindsay/Documents/data/Rondeau/tmp2.dep' --flat_increment=0.001 -v
```
<img width="965" alt="Screen Shot 2019-03-10 at 9 34 22 AM" src="https://user-images.githubusercontent.com/8441542/54085784-101b2000-4318-11e9-9203-3c65f04441f7.png">

Constrained-breached DEM:

```
./whitebox_tools --run="Breachdepressions" --wd="/Users/johnlindsay/Documents/data/Rondeau" --dem='/Users/johnlindsay/Documents/data/Rondeau/DEM_clipped.dep' --output='/Users/johnlindsay/Documents/data/Rondeau/tmp1.dep' --max_depth=10 --max_length=70 --flat_increment=0.001 -v
```
<img width="966" alt="Screen Shot 2019-03-10 at 9 34 35 AM" src="https://user-images.githubusercontent.com/8441542/54085804-45277280-4318-11e9-8585-3161a71d0f11.png">

**Comment by spaceof7** [comment]
At first I was using the QGIS plugin but when it didnt work I started running WhiteboxTools at the command line.  Initially I was specifying max_depth and max_length but also tested it unconstrained.   Oh, also I was using the tools on .tif files, but also tested on a .dep version. 

I noticed that in both examples you specified a flat_increment so i tested using the same parameters you used and it worked as expected.  I tested a few variations of parameters and it seems that if there is no flat_increment parameter specified then the ouput is unmodified.

Also, I wanted to say I really appreciate WhiteboxTools and not having to do file conversion/software switching to get the data in and out of Whitebox GAT.  Thanks!

**Comment by jblindsay** [comment]
That was a good spot @spaceof7. I just confirmed your finding; it would seem that by not specifying the flat_increment value, the output looks the same as the input. I'll spend a bit of time now digging into why this odd behaviour occurs.

**Comment by jblindsay** [comment]
Okay, I found the error. Because the tool is intended to automatically calculate an appropriate value of the flat increment, based on the input DEM elevation range, in the case that this parameter is not specified by the user, the value was initialized to NaN. 

```
        let mut flat_increment = f64::NAN;
```

Later in the code, I tested to see if the value of the variable was NaN, in which case the tool was supposed to calculate the value automatically:

```
       let small_num = if flat_increment != f64::NAN {
            flat_increment
        } else {
            let min_val = input.configs.minimum;
            let elev_digits = ((input.configs.maximum - min_val) as i64).to_string().len();
            let elev_multiplier = 10.0_f64.powi((5 - elev_digits) as i32);
            1.0_f64 / elev_multiplier as f64
       };
```
 
The problem was that NaN != NaN in Rust. And so I needed to modify the line to read:

```
        let small_num = if !flat_increment.is_nan() {
```

This was a very subtle bug that I was unlikely to ever find had you not reported the issue. As such, I truly thank you for taking the time. The updated code has been submitted to the repository and the fix to the tool will be released with the next version (likely 0.15.1). I'm closing this issue as resolved.
